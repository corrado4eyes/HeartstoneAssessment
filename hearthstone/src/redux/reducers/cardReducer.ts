import {CardActionsType, CardActions, CardFilters, OnSubmitFilterAction} from '../actions/cardActions';
import Card from '../../model/card';
import { CardSet } from '../../model/cardSet';
import { Rarity } from '../../model/rarity';
import { CardClass } from '../../model/class';
import { CardType } from '../../model/cardType';
import { Mechanic } from '../../model/mechanic';

export interface State {
    cards: Card[]
    filteredCards: Card[]
    loading: boolean
    error: string
    filters: {
        [key: string]: string
    }
}

export const initialState = {
    cards: [],
    loading: false,
    error: "",
    filters: {
        cardSet: CardSet.Basic,
        rarity: Rarity.Free,
        playerClass: CardClass.Druid,
        type: CardType.Hero,
        mechanics: Mechanic.Taunt
    },
    filteredCards: []
}

const updateCards = (array: Card[], updatedCard: Card) => {
    return array.map((card: Card) => {
        if(card.cardId! === updatedCard.cardId!)
            return updatedCard
        else
            return card
    }) 
}
const filterMechanics = (cards: Card[], action: OnSubmitFilterAction) => {
    let data: Card[] = []
    cards = cards.filter((el: Card) => el.mechanics)
        cards.forEach((el: Card) => {
            if(el.mechanics) {
                el.mechanics.map((mechanic: any) => {
                    if(mechanic.name === action.filter) {
                        data.push(el)
                    }
                })
            }
        });
        return data
}

const filterCards = (action: OnSubmitFilterAction, cards: Card[], filter: any): Card[] => {
    
    if(action.filterKey === "mechanics"){
        return filterMechanics(cards, action)
    } else {
        return cards.filter((el: Card) => (el as any)[action.filterKey] === filter[action.filterKey]);
    }
}

export const reducer = (state: State = initialState, action: CardActionsType) => {
    const stateCopy = Object.assign({}, state)
    let filteredCards: Card[] = []
    switch(action.type) {
        case CardActions.onSyncCards:
            return Object.assign({}, state, {
                loading: true
            });
        case CardActions.onSyncCardsSucceed:
        const cardState = stateCopy.cards.concat(action.cards)
        const filteredCardState = cardState.filter((el: Card) => el.cardSet === state.filters['cardSet'])
            return Object.assign({}, state, {
                loading: false,
                cards: cardState,
                filteredCards: filteredCardState
            });
        case CardActions.onSyncCardsFailed:
            return Object.assign({}, state, {
                loading: false,
                error: action.error,
            });
        case CardActions.onSubmitFilter: 
            const newFilter = Object.assign({}, stateCopy.filters, {[action.filterKey]: action.filter});
            filteredCards = filterCards(action, stateCopy.cards, newFilter);
            return Object.assign({}, state, {
                filteredCards: filteredCards,
                filters: newFilter
            });
        case CardActions.onFilterByName:
            filteredCards = stateCopy.cards.filter((el: Card) => {
                if (el.name!.indexOf(action.name) >= 0) {
                    return el
                }else{}
            });
            return Object.assign({}, state, {
                filteredCards
            });
        case CardActions.onSaveCardSucceed:
            const cards = updateCards(stateCopy.cards, action.card)
            filteredCards = updateCards(stateCopy.filteredCards, action.card)
            return Object.assign({}, stateCopy, {
                cards: cards,
                filteredCards: filteredCards
            });
        case CardActions.onSaveCardFailed:
            return Object.assign({}, stateCopy, {
                error: action.error,
            })
        default:
            return Object.assign({}, state);
    }
}