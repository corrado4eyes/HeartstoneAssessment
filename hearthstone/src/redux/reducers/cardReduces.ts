import {CardActionsType, CardActions} from '../actions/cardActions';
import Card from '../../model/card';

export interface State {
    cards: Card[]
    filteredCards: Card[]
    loading: boolean
    error: string
    filters: {
        [key: string]: string
    }
}

export const initialState = {
    cards: [],
    loading: false,
    error: "",
    filters: {
        cardSet: "Basic",
        rarity: "Free",
    },
    filteredCards: []
}

export const reducer = (state: State = initialState, action: CardActionsType) => {
    const stateCopy = Object.assign({}, state)
    let filteredCards: Card[] = []
    switch(action.type) {
        case CardActions.onSyncCards:
            return Object.assign({}, state, {
                loading: true
            });
        case CardActions.onSyncCardsSucceed:
        const cardInitialState = action.cards.filter((el: Card) => el.cardSet === state.filters['cardSet'])
            return Object.assign({}, state, {
                loading: false,
                cards: action.cards,
                filteredCards: cardInitialState
            });
        case CardActions.onSyncCardsFailed:
            return Object.assign({}, state, {
                loading: false,
                error: action.error,
            });
        case CardActions.onSubmitFilter: 
            const newFilter = Object.assign({}, stateCopy.filters, {[action.filterKey]: action.filter});
            filteredCards = stateCopy.cards.filter((el: Card) => (el as any)[action.filterKey] === newFilter[action.filterKey]);
            return Object.assign({}, state, {
                filteredCards: filteredCards,
                filters: newFilter
            });
        case CardActions.onFilterByName:
            filteredCards = stateCopy.cards.filter((el: Card) => {
                if (el.name!.indexOf(action.name) >= 0) {
                    return el
                }else{}
            });
            return Object.assign({}, state, {
                filteredCards
            });
        default:
            return Object.assign({}, state);
    }
}