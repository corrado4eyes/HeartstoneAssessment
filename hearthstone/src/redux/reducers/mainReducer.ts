import {combineReducers} from 'redux';
import * as fromCard from './cardReducer';
export interface RootState {
    card: fromCard.State
}

export const initialState: RootState = {
    card: fromCard.initialState
}

export const mainReducer = combineReducers<RootState>({card: fromCard.reducer}) 