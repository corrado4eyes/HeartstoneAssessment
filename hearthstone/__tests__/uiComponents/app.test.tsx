import React from 'react';
import ConnectedApp, {App} from '../../src/uiComponents/app';
import {shallow, mount, ShallowWrapper, ReactWrapper} from 'enzyme'
import { dummyCardArray } from '../__mocks__/mockObjects';
import Card from '../../src/model/card';
import { RootState, initialState } from '../../src/redux/reducers/mainReducer';
import { Provider } from 'react-redux';
import { generateMockStore } from '../__mocks__/mockStore';

describe('App', () => {
    const cardSet = "Dummy"
    describe('App(unconnected)', () => {
        const dispatchSyncCard = jest.fn()
        const cards: Card[] = dummyCardArray
        const component: ShallowWrapper = shallow(<App cardSet={cardSet} cards={cards} dispatchSyncCard={dispatchSyncCard}/>);
        it('renders n elements in an array', () => {
            component.setState({cards: cards})
            const divs: ShallowWrapper = component.find('.card');
            expect(divs).toHaveLength(cards.length);
        });
    
        it('Snapshot testing', () => {
                component.setState({cards: cards})
                expect(component).toMatchSnapshot();
        });

        it('fires dispatchSyncCard onComponentDidMount', () => {
            expect(dispatchSyncCard).toHaveBeenCalledTimes(1);
        });
    });

    describe('App(connected)', () => {
        const state: RootState = {
            ...initialState, 
        }
        state.card.cards = dummyCardArray;
        state.card.cardSet = cardSet
        const store = generateMockStore(state)
        const component: ReactWrapper = mount(<Provider store={store}><ConnectedApp cardSet={cardSet}/></Provider>);
        it('checks the props', () => {
            expect(component.find(App).prop('cards')).toBe(dummyCardArray)
        });
    });
});