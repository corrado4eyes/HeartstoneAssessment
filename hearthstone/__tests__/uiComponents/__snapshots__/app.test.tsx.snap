// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App(unconnected) Snapshot testing 1`] = `
<Container
  fluid={false}
>
  <ForwardRef
    noGutters={false}
  >
    <Connect(FilterBar) />
  </ForwardRef>
  <ForwardRef
    noGutters={false}
  >
    <Col
      className="card no-border"
      key="what a dummy card!"
      md="4"
    >
      <CardComponent
        card={
          Object {
            "artist": "Corrado Quattrocchi",
            "cardId": "what a dummy card!",
            "cardSet": "Dummy",
            "collectible": true,
            "cost": 9999,
            "faction": "Splendo",
            "flavor": "Meat",
            "howToGet": "Ask Frits",
            "howToGetGold": "Feeds and Walk Frits for 2 years every day (excluded the holidays)",
            "img": "http://media.services.zam.com/v1/media/byName/hs/cards/enus/DS1_183.png",
            "imgGold": "http://media.services.zam.com/v1/media/byName/hs/cards/enus/animated/DS1_183_premium.gif",
            "locale": "enUS",
            "name": "corrado4eyes",
            "playerClass": "Human",
            "rarity": "Super Rare",
            "text": "A dummy Italian boy",
            "type": "Developer",
          }
        }
      />
    </Col>
    <Col
      className="card no-border"
      key="what an ugly dummy card!"
      md="4"
    >
      <CardComponent
        card={
          Object {
            "artist": "lala",
            "cardId": "what an ugly dummy card!",
            "cardSet": "Not Dummy",
            "collectible": true,
            "cost": 9999,
            "faction": "Splendo",
            "flavor": "Meat",
            "howToGet": "Ask Frits",
            "howToGetGold": "Feeds and Walk Frits for 2 years every day (excluded the holidays)",
            "img": "http://media.services.zam.com/v1/media/byName/hs/cards/enus/DS1_183.png",
            "imgGold": "http://media.services.zam.com/v1/media/byName/hs/cards/enus/animated/DS1_183_premium.gif",
            "locale": "enUS",
            "name": "lalala",
            "playerClass": "Human",
            "rarity": "Not Super Rare",
            "text": "A dummy Italian boy",
            "type": "Developer",
          }
        }
      />
    </Col>
  </ForwardRef>
</Container>
`;
